libraries/chain/steem_evaluator.cpp

```diff
@@ -2232,89 +2231,67 @@ namespace golos { namespace chain {
-                    "Account does not have enough vesting shares to delegate.");
+            FC_ASSERT((increasing ? delta : -delta) >= min_update,
+                "Delegation difference is not enough. min_update: ${min}", ("min", min_update));
+#ifdef STEEMIT_BUILD_TESTNET
+            // min_update depends on account_creation_fee, which can be 0 on testnet
+            FC_ASSERT(delta.amount != 0, "Delegation difference can't be 0");
+#endif
 
+            if (increasing) {
+                FC_ASSERT(delegator.available_vesting_shares(true) >= delta,
+                    "Account does not have enough vesting shares to delegate.");
+                if (!delegation) {
+                    FC_ASSERT(op.vesting_shares >= min_delegation,
+                        "Account must delegate a minimum of ${v}", ("v",min_delegation));
+                    _db.create<vesting_delegation_object>([&](vesting_delegation_object& o) {
+                        o.delegator = op.delegator;
+                        o.delegatee = op.delegatee;
+                        o.vesting_shares = op.vesting_shares;
+                        o.min_delegation_time = _db.head_block_time();
+                    });
+                }
                 _db.modify(delegator, [&](account_object& a) {
                     a.delegated_vesting_shares += delta;
                 });
-                _db.modify(delegatee, [&](account_object& a) {
-                    a.received_vesting_shares += delta;
-                });
-                _db.modify(*delegation, [&](vesting_delegation_object& o) {
-                    o.vesting_shares = op.vesting_shares;
-                });
-            } else { /* delegation->vesting_shares > op.vesting_shares */
-            // Delegation is decreasing
-                auto delta = delegation->vesting_shares - op.vesting_shares;
-                if (op.vesting_shares.amount > 0) {
-                    FC_ASSERT(delta >= min_update,
-                        "Golos Power decrease is not enough of a difference. min_update: ${min}", ("min", min_update));
-                    FC_ASSERT(op.vesting_shares >= min_delegation,
-                        "Delegation must be removed or leave minimum delegation amount of ${v}", ("v", min_delegation));
-                } else {
-                    FC_ASSERT(delegation->vesting_shares.amount > 0,
-                        "Delegation would set vesting_shares to zero, but it is already zero");
-                }
+            } else {
+                FC_ASSERT(op.vesting_shares.amount == 0 || op.vesting_shares >= min_delegation,
+                    "Delegation must be removed or leave minimum delegation amount of ${v}", ("v", min_delegation));
                 _db.create<vesting_delegation_expiration_object>([&](vesting_delegation_expiration_object& o) {
                     o.delegator = op.delegator;
-                    o.vesting_shares = delta;
+                    o.vesting_shares = -delta;
```

Comment:
Why it is negative?

Reply:
`delta` itself is negative here, put to operation positive value.
And it needs to be negative at:
```
 _db.modify(delegatee, [&](account_object& a) {
    a.received_vesting_shares += delta;
});
```
which is universal for increasing/decreasing delegation

Reply:
Alright. Another question for this particular case in here: https://github.com/GolosChain/golos/pull/585/commits/e685658ff45aacd9d2431fb1e84d7779abe8fe5e

Why don't you just copy the Steem Power delegation from steem? I mean, that is the trivial logic and it does not require to be reviewed.

Reply:
My fault. It is fine.

Reply:
My implementation is trivial too, it's even simpler than Steem's one. Just removed duplicated code from branches and put it after assertions and unique logic of each branch. 
Pseudocode is less than 15 lines of code:
```
delegation = find_existing_delegation
if increasing_delegation {
    if !delegation
        create_delegation
    delegator.delegated_vesting_shares += delta
} else
     create vesting_delegation_expiration_object

delegatee.received_vesting_shares += delta
if delegation {
    if op.vesting_shares.amount > 0
        delegation.vesting_shares = op.vesting_shares
    else
        remove delegation
}
```

Steem implementation have lot of duplicated code in each branch, which can be costly to maintain.